fix(config): properly apply defaults when reading folder configuration

pixelspark opened the pull request:
### Purpose

This fixes `File system type '' not recognized`. 

Previously, the FolderConfiguration object would be zero-initialized (and filesystemType = 0 = FileSystemTypeBasic). Now, with FileSystemType being essentially a string, it is set to the empty string on initialization. There was a `default:"basic"` annotation on the field, but that default is only applied when calling `structutil.SetDefaults` manually. This was not done for FolderConfiguration (it was implemented for VersioningConfiguration). This PR adds `UnmarshalXML` and `UnmarshalJSON` implementations for FolderConfiguration that call `SetDefaults` before applying values from the XML/JSON, fixing this issue and making `basic` the default FilesystemType again when none is specified.

### Testing

Remove `<filesystemType>basic</filesystemType>` from configuration, then start Syncthing. It should work the same.

### Screenshots

n/a

### Documentation

n/a (This PR makes Syncthing configuration work [according to the documentation](https://docs.syncthing.net/advanced/folder-filesystem-type.html) again). 

## Authorship

Your name and email will be added automatically to the AUTHORS file
based on the commit metadata.

pixelspark committed:
fix(config): properly apply defaults when reading folder configuration

pixelspark committed:
chore(config): folder path should not default to "~"

According to documentation, the path field is mandatory and should not have a default. This is also what should be expected in TestDefaultValues.

pixelspark commented:
OK, so figuring out why the tests failed took a bit more time. It seems there was some misalignment between the defaults set on the FolderConfiguration struct, what `FolderConfiguration.prepare` is doing to that, and what the test is checking. See the `Path` field in particular (according to docs it is mandatory and has no default, in the struct it said `default:"~"` but then `TestDefaultValues` was testing for getting an error because of an empty path when none specified).

tomasz1986 commented on lib/config/config_test.go:
```diff
@@ -173,33 +174,28 @@ func TestDeviceConfig(t *testing.T) {
+		fc.AutoNormalize = true
+		fc.MinDiskFree = Size{1, "%"}
+		fc.MaxConflicts = -1
+		fc.Versioning.CleanupIntervalS = 1
```
Is this on purpose? There is also `fc.Versioning.CleanupIntervalS = 3600` a few lines below.

pixelspark replied:
Right, should just set CleanupIntervalS once!

calmh commented on lib/config/config_test.go:
```diff
@@ -173,33 +174,27 @@ func TestDeviceConfig(t *testing.T) {
-				},
-			},
-		}
+		fc := FolderConfiguration{}
+		structutil.SetDefaults(&fc)
```
We don't want to do this here. The purpose of the test, such as it is, is to verify that folders are created with the default values we expect. Running the default-setter on the expected result makes it an `a == a` type of comparison, so I think we want to keep this manually defined.

calmh commented on lib/config/config_test.go:
```diff
@@ -219,7 +214,18 @@ func TestDeviceConfig(t *testing.T) {
+			device4: &expectedDevices[1],
+		}
+
+		fc.prepare(device1, expectedDeviceConfigs) // Called by ReadXML, so we should call it too
```
Same

pixelspark commented:
Right, the intent of the changes to the test was to stay close to the original test (which was not explicit about default values either, instead relying on zero initialization of fields which *happen to be* the default value, so implicitly it was complete but also inconsistent with the default values set for `structutil`'s use). I will add all fields and their expected values here. 

OT, is there a way to tell Go to require initialization of all struct fields in specific cases, so we remember to update tests like this on changes? This is one of the less ideal design choices in Go imho.

calmh commented:
No, but zero valued fields are irrelevant here, and don't require updating the test.

pixelspark committed:
chore(config): fix TestDeviceConfig

Because we now actually apply structutil defaults for FolderConfiguration fields on UnmarshalXML,
the test now sees these defaults instead of the (Go-default) zero values for these fields when
they are not set from the XML file that is read.

calmh committed:
Merge branch 'main' into fix/default-fstype

baekelite commented:
I've noticed that on a fresh installation of 1.29.6, the default folder path (for new folders, or shared folder) is now blank, while it has been the home directory for as long as I can remember.

Believe this change did it: https://github.com/syncthing/syncthing/commit/f7c8efd93c49ad0bb865b117b5b3d244622a5712#diff-e2226aaf2241995d4407eb03a9516663bd730d7331626efa5e80f89c6dd90621L51

Is this change intended?  As a user, this change was unexpected.
